VIMS-389 Support updating and getting VVM status for receiving stock

chunky56 opened the pull request:
VVM status needs to be updated when receiving stock. VVM status is stored per stock card entry. To make this global, need to create new table for key-value pairs (stock_card_entry_key_values) and set vvmstatus as a key-value. Mappers to get key-values returns a list, rather than a map, because of limitations of MyBatis.

Also, create new stock event types to support issue, receipt, adjustments. Change controller code to support these new types. Only set VVM status if found in the event and type is receipt.

chunky56 committed:
VIMS-389 Support updating and getting VVM status for receiving stock

VVM status needs to be updated when receiving stock. VVM status is stored per stock card entry. To make this global, need to create new table for key-value pairs (stock_card_entry_key_values) and set vvmstatus as a key-value. Mappers to get key-values returns a list, rather than a map, because of limitations of MyBatis.

Also, create new stock event types to support issue, receipt, adjustments. Change controller code to support these new types. Only set VVM status if found in the event and type is receipt.

benleibert commented on src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java:
```diff
@@ -158,9 +159,9 @@ public ResponseEntity getStockCards(@PathVariable Long facilityId,
-    public ResponseEntity adjustStock(@PathVariable long facilityId,
-                                      @RequestBody(required=true) List<StockEvent> events,
-                                      HttpServletRequest request) {
+    public ResponseEntity processStock(@PathVariable long facilityId,
```
The processStock method relies on a private method called getLotOnHand, and defined in this controller. I wonder, though, whether the getLotOnHand method better belongs in the service-layer. It has a fair amount of logic in it – perhaps more than belongs at the controller level.

Also, the getLotOnHand method makes interesting use of a StringBuilder to return errors to the user. Is this done elsewhere in the System? At first glance, a developer might think a StringBuilder is used rather than Java’s inbuilt try-catch mechanism because we want the ability to build and return a list of multiple errors. That’s not the case in this instance, however, so I wonder if the standard try-catch approach may be preferable.

chunky56 committed:
Make code to set vvmStatus more generic

Right now, StockEvent and API in controller has code to set vvmStatus, but it needs to be more generic to support any custom properties in the StockEvent.

benleibert commented on src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java:
```diff
@@ -183,17 +186,20 @@ public ResponseEntity adjustStock(@PathVariable long facilityId,
             if(null == card)
-                return OpenLmisResponse.error("Unable to adjust stock for facility and product",
+                return OpenLmisResponse.error("Unable to process stock for facility and product",
                     HttpStatus.BAD_REQUEST);
```
I don’t think this is a BAD_REQUEST. Rather, a null StockCard at this point indicates a server-side error (code 500).

benleibert commented on src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java:
```diff
@@ -205,12 +211,30 @@ public ResponseEntity adjustStock(@PathVariable long facilityId,
-            StockCardEntry entry = new StockCardEntry(card,
-                StockCardEntryType.ADJUSTMENT,
-                quantity);
+            if (null != reason) {
```
StockEvent.getQuantity() already has a bit of logic related to the sign (positive or negative) of the value it returns. The rest of this logic is generic and useful enough that it probably belongs in StockEvent as well.

chunky56 committed:
Adjust json object graph to hide isValid and show keyValues

chunky56 committed:
Make improvements based on code review

* Move getLotOnHand to service
* Move quantity adjustment code to StockEvent class
* Move tests
* Fix NPE bug
